### PR TITLE
[DRAFT] Added Lightbulb service to Fanv2

### PR DESCRIPTION
**Pull request by slavvka**
I added Lightbulb service to Fanv2 and it almost works. The only issue is that with every consequent change of Lightbulb in HomeKit the number of requests to Tuya doubles and it starts to fail with an error:
```
This callback function has already been called by someone else; it can only be called one time.
```

Full log for the second change (two requests instead of one for light and brightness):
```
TuyaOpenAPI request: method = post, endpoint = https://openapi.tuyaus.com, path = /v1.0/devices/850007548caab5ebd0bc/commands, params = null, body = {"commands":[{"code":"bright_value","value":326}]}, headers = {"t":"1628100416778","client_id":"***","nonce":"***","Signature-Headers":"client_id","sign":"***","sign_method":"HMAC-SHA256","access_token":"***","lang":"en","dev_lang":"javascript","dev_channel":"homebridge","devVersion":"1.3.1"}
TuyaOpenAPI request: method = post, endpoint = https://openapi.tuyaus.com, path = /v1.0/devices/850007548caab5ebd0bc/commands, params = null, body = {"commands":[{"code":"bright_value","value":326}]}, headers = {"t":"1628100416779","client_id":"***","nonce":"***","Signature-Headers":"client_id","sign":"***","sign_method":"HMAC-SHA256","access_token":"***","lang":"en","dev_lang":"javascript","dev_channel":"homebridge","devVersion":"1.3.1"}
TuyaOpenAPI request: method = post, endpoint = https://openapi.tuyaus.com, path = /v1.0/devices/850007548caab5ebd0bc/commands, params = null, body = {"commands":[{"code":"light","value":true}]}, headers = {"t":"1628100416779","client_id":"***","nonce":"***","Signature-Headers":"client_id","sign":"***","sign_method":"HMAC-SHA256","access_token":"***","lang":"en","dev_lang":"javascript","dev_channel":"homebridge","devVersion":"1.3.1"}
TuyaOpenAPI request: method = post, endpoint = https://openapi.tuyaus.com, path = /v1.0/devices/850007548caab5ebd0bc/commands, params = null, body = {"commands":[{"code":"light","value":true}]}, headers = {"t":"1628100416780","client_id":"***","nonce":"***","Signature-Headers":"client_id","sign":"***","sign_method":"HMAC-SHA256","access_token":"***","lang":"en","dev_lang":"javascript","dev_channel":"homebridge","devVersion":"1.3.1"}
TuyaOpenAPI response: {"result":true,"success":true,"t":1628100417076} path = /v1.0/devices/850007548caab5ebd0bc/commands
TuyaOpenAPI response: {"result":true,"success":true,"t":1628100417078} path = /v1.0/devices/850007548caab5ebd0bc/commands
TuyaOpenAPI response: {"result":true,"success":true,"t":1628100417080} path = /v1.0/devices/850007548caab5ebd0bc/commands
[SET][Master Bedroom Lights & Fan] Characteristic Error: Error: This callback function has already been called by someone else; it can only be called one time.
    at /usr/local/lib/node_modules/homebridge/node_modules/hap-nodejs/src/lib/util/once.ts:6:13
    at /Users/sviatoslavmankivskyi/Projects/tuya-homebridge/lib/fanv2_accessory.js:155:11
    at processTicksAndRejections (internal/process/task_queues.js:93:5)
(node:23763) UnhandledPromiseRejectionWarning: Error: This callback function has already been called by someone else; it can only be called one time.
    at /usr/local/lib/node_modules/homebridge/node_modules/hap-nodejs/src/lib/util/once.ts:6:13
    at /Users/sviatoslavmankivskyi/Projects/tuya-homebridge/lib/fanv2_accessory.js:159:11
    at processTicksAndRejections (internal/process/task_queues.js:93:5)
(Use `node --trace-warnings ...` to show where the warning was created)
(node:23763) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 1)
(node:23763) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
TuyaOpenAPI response: {"result":true,"success":true,"t":1628100417098} path = /v1.0/devices/850007548caab5ebd0bc/commands
[SET][Master Bedroom Lights & Fan] Characteristic Error: Error: This callback function has already been called by someone else; it can only be called one time.
    at /usr/local/lib/node_modules/homebridge/node_modules/hap-nodejs/src/lib/util/once.ts:6:13
    at /Users/sviatoslavmankivskyi/Projects/tuya-homebridge/lib/fanv2_accessory.js:155:11
    at processTicksAndRejections (internal/process/task_queues.js:93:5)
(node:23763) UnhandledPromiseRejectionWarning: Error: This callback function has already been called by someone else; it can only be called one time.
    at /usr/local/lib/node_modules/homebridge/node_modules/hap-nodejs/src/lib/util/once.ts:6:13
    at /Users/sviatoslavmankivskyi/Projects/tuya-homebridge/lib/fanv2_accessory.js:159:11
    at processTicksAndRejections (internal/process/task_queues.js:93:5)
(node:23763) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 2)
TuyaOpenMQ onMessage: topic = cloud/token/in/fcc6352002be94a08fe364b7f5ab439b, message = {"data":{"dataId":"6deafac7-2916-47cc-bcab-13ec573f39ed","devId":"850007548caab5ebd0bc","productKey":"ipyqwvhvzzjezv35","status":[{"9":true,"code":"light","t":"1628100416","value":true}]},"protocol":4,"pv":"2.0","sign":"***","t":1628100416}
TuyaOpenMQ onMessage: topic = cloud/token/in/fcc6352002be94a08fe364b7f5ab439b, message = {"data":{"dataId":"1411a283-56b4-4628-9e0b-abff359a12e0","devId":"850007548caab5ebd0bc","productKey":"ipyqwvhvzzjezv35","status":[{"10":326,"code":"bright_value","t":"1628100416","value":326}]},"protocol":4,"pv":"2.0","sign":"***","t":1628100416}
TuyaOpenMQ onMessage: topic = cloud/token/in/fcc6352002be94a08fe364b7f5ab439b, message = {"data":{"dataId":"133a5692-88df-4c6a-b899-33cff32995b2","devId":"850007548caab5ebd0bc","productKey":"ipyqwvhvzzjezv35","status":[{"9":true,"code":"light","t":"1628100416","value":true}]},"protocol":4,"pv":"2.0","sign":"***","t":1628100416}
TuyaOpenMQ onMessage: topic = cloud/token/in/fcc6352002be94a08fe364b7f5ab439b, message = {"data":{"dataId":"f5ce9721-2e94-41cd-82af-0f913ad87344","devId":"850007548caab5ebd0bc","productKey":"ipyqwvhvzzjezv35","status":[{"10":326,"code":"bright_value","t":"1628100416","value":326}]},"protocol":4,"pv":"2.0","sign":"***","t":1628100416}
```

I have no ideas left what could be the reason. Please help!

### CODE REVIEW

**Comment by slavvka** [comment]
@Hanh94 Could you please take a look at the approach here? Also you may have an idea what's wrong with it?

File: lib/fanv2_accessory.js [hunk 1]
```diff
@@ -76,23 +110,36 @@ class Fanv2Accessory extends BaseAccessory {
         const hbSwing = this.tuyaParamToHomeBridge(Characteristic.SwingMode, this.swingMap.value);
         this.normalAsync(Characteristic.SwingMode, hbSwing, this.isRefresh)
       }
+
+      if (this.lightService && statusMap.code === 'light') {
+        this.switchLed = statusMap;
+        const hbLight = this.tuyaParamToHomeBridge(Characteristic.On, this.switchLed.value);
+        this.normalAsync(Characteristic.On, hbLight, this.isRefesh, this.lightService)
```
Comment (Hanh94) [inline lib/fanv2_accessory.js:121]: `this.normalAsync(Characteristic.On, hbLight, this.isRefesh, this.lightService)` , Causes Characteristic to be created after each MQTT callback ,should use **this.isRefresh**

File: lib/fanv2_accessory.js [hunk 2]
```diff
@@ -76,23 +110,36 @@ class Fanv2Accessory extends BaseAccessory {
...
+      }
+
+      if (this.lightService && statusMap.code === 'bright_value') {
+        this.brightValue = statusMap;
+        const rawValue = this.brightValue.value;
+        const percentage = Math.floor((rawValue - this.bright_range.min) * 100 / (this.bright_range.max - this.bright_range.min)); //    $
+        this.normalAsync(Characteristic.Brightness, percentage > 100 ? 100 : percentage, this.isRefesh, this.lightService)
```
Comment (Hanh94) [inline lib/fanv2_accessory.js:128]: same too

**Review by Hanh94 — COMMENTED** [review]
Hi @slavvka ,Thank you for your PR.
We have reviewed your code and will merge the code into the master branch. If necessary, we will adjust the details and  hope you don't mind. Thanks again.

**Pull request merged**